File: src/main/java/arcana/common/packets/MarksPacket.java
package arcana.common.packets;

import arcana.common.aspects.AspectUtils;
import arcana.common.capability.Marks;
import net.minecraft.client.Minecraft;
import net.minecraft.client.world.ClientWorld;
import net.minecraft.network.PacketBuffer;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.fml.network.NetworkDirection;
import net.minecraftforge.fml.network.NetworkEvent;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.function.Supplier;
import java.util.stream.Collectors;

public class MarksPacket extends PacketHandler<MarksPacket> {
    Marks cap;
    public MarksPacket(Marks cap){
        this.cap = cap;
    }
    public MarksPacket(){}

    public MarksPacket decode(PacketBuffer buf){
        Marks cap = new Marks(new ArrayList[AspectUtils.primalAspects.length]);
        for (int i = 0; i < AspectUtils.primalAspects.length; i++) {
            long[] arr = buf.readLongArray(null);
            cap.positions[i] = Arrays.stream(arr).mapToObj(BlockPos::of).collect(Collectors.toList());
        }
        return new MarksPacket(cap);
    }

    public void encode(MarksPacket packet, PacketBuffer buf){
        for (int i = 0; i < AspectUtils.primalAspects.length; i++) {
            long[] arr = packet.cap.positions[i].stream().mapToLong(BlockPos::asLong).toArray();
            buf.writeLongArray(arr);
        }
    }

    public void innerHandle(MarksPacket packet, Supplier<NetworkEvent.Context> ctx){
        if (ctx.get().getDirection() == NetworkDirection.PLAY_TO_CLIENT) {
            ClientWorld world = Minecraft.getInstance().level;
            if (world == null)
                return;
            Marks clientCap = world.getCapability(Marks.CAPABILITY).resolve().orElse(null);
            clientCap.setNbt(packet.cap.getNbt());
        } else if (ctx.get().getDirection() == NetworkDirection.PLAY_TO_SERVER){
            //not required
        }
    }
}

File: src/main/kotlin/arcana/common/packets/MarksPacket.kt
package arcana.common.packets

import arcana.common.aspects.Aspects
import arcana.common.capability.Marks
import arcana.common.capability.getMarks
import net.minecraft.client.Minecraft
import net.minecraft.network.PacketBuffer
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.network.NetworkDirection
import net.minecraftforge.fml.network.NetworkEvent
import java.util.function.Supplier

class MarksPacket(var cap: Marks? = null)
    : PacketHandler<MarksPacket>() {
    override fun decode(buf: PacketBuffer): MarksPacket {
        val cap = Marks(Array(Aspects.primal.size) {
            val arr = buf.readLongArray(null)
            arr.map{pPackedPos: Long -> BlockPos.of(pPackedPos)}.toMutableList()
        })
        return MarksPacket(cap)
    }

    override fun encode(packet: MarksPacket, buf: PacketBuffer) {
        for (i in Aspects.primal.indices) {
            val arr = packet.cap!!.positions[i].stream().mapToLong { obj: BlockPos? -> obj!!.asLong() }.toArray()
            buf.writeLongArray(arr)
        }
    }

    override fun innerHandle(packet: MarksPacket, ctx: Supplier<NetworkEvent.Context>) {
        if (ctx.get().direction == NetworkDirection.PLAY_TO_CLIENT) {
            val world = Minecraft.getInstance().level ?: return
            val clientCap = world.getMarks()
            clientCap.nbt = packet.cap!!.nbt
        } else if (ctx.get().direction == NetworkDirection.PLAY_TO_SERVER) {
            //not required
        }
    }
}